### ISSUE TITLE
Celery asyncio tasks

### ISSUE THREAD

**Issue by txomon** [celery/celery#3884]
Hello! 

I have found that there is no integration of celery with asyncio, I expected to have a way to run the workers in an ioloop so that I could execute several IOBound tasks at the same time (I already have all the business logic implemented in asyncio), but I found that the only supported integration is eventlet.

Also, I haven't found any way to execute more than one task at a time in a worker. (use prefetch_count >1)

## Expected behavior
To be able to write `async` tasks

## Actual behavior
I can only patch all the IO in python to get async

**Comment by thedrow** [comment]
Celery 5.0 will support asyncio. We currently do not support it.

**Comment by dvarrazzo** [comment]
It doesn't seem Celery 5.0 support async that much.

https://docs.celeryproject.org/en/stable/search.html?q=asyncio

**Comment by thedrow** [comment]
We planned to do it but elected to go with a few major releases with a smaller scope.
I'll update everyone when it comes.

**Comment by spawn-guy** [comment]
@thedrow any news on official `asyncio` tasks/pools? it's been a while

**Comment by devraj** [comment]
@spawn-guy @dvarrazzo have you guys managed to work around this at all? I am in a fix where I migrated our new stack with SQLAlchemy and FastAPI  and everything else about to use asyncio, alas I hadn't checked that Celery does not have support for it, although v5 has been out for a while.

My preference would be to find a workaround to be able to use the async code in sync mode and then move to asyncio when the project supports it.

Thanks for any insights.

**Comment by auvipy** [comment]
> @spawn-guy @dvarrazzo have you guys managed to work around this at all? I am in a fix where I migrated our new stack with SQLAlchemy and FastAPI and everything else about to use asyncio, alas I hadn't checked that Celery does not have support for it, although v5 has been out for a while.
> 
> My preference would be to find a workaround to be able to use the async code in sync mode and then move to asyncio when the project supports it.
> 
> Thanks for any insights.

celery 6 will be the new target, and main major change of it will be async await compatibility. in the mean time you code try django projects https://github.com/django/asgiref Sync-to-async and async-to-sync function wrappers, asgiref.sync

**Comment by spawn-guy** [comment]
@devraj we are successfully using https://github.com/kai3341/celery-pool-asyncio for a few years now. not with `sqlalchemy` but with raw `asyncpg`/`cassandra` and `fastapi`

here is minimal example repo (not mine, i just googled it) https://github.com/codesutras/celery-pool-asyncio-example - the config we use is almost the same.

there are some quirks to use the `celery_app.send_task()` as in debug mode i don't use the pool(so no `await` keyword), but in production we do (`await` keyword is needed). this can be fixed with some extra configuration, i guess.


@auvipy awesome news! i hope this time it will be merged and done in `v6` (i remember promises over `v5` inclusion)

**Comment by auvipy** [comment]
> @devraj we are successfully using https://github.com/kai3341/celery-pool-asyncio for a few years now. not with `sqlalchemy` but with raw `asyncpg`/`cassandra` and `fastapi`
> 
> here is minimal example repo (not mine, i just googled it) https://github.com/codesutras/celery-pool-asyncio-example - the config we use is almost the same.
> 
> there are some quirks to use the `celery_app.send_task()` as in debug mode i don't use the pool(so no `await` keyword), but in production we do (`await` keyword is needed). this can be fixed with some extra configuration, i guess.
> 
> @auvipy awesome news! i hope this time it will be merged and done in `v6` (i remember promises over `v5` inclusion)

I get very few money to work on celery right now. it is a huge overhaul work. take lots of time and technical difficulties

**Comment by devraj** [comment]
> > @devraj we are successfully using https://github.com/kai3341/celery-pool-asyncio for a few years now. not with `sqlalchemy` but with raw `asyncpg`/`cassandra` and `fastapi`
> > here is minimal example repo (not mine, i just googled it) https://github.com/codesutras/celery-pool-asyncio-example - the config we use is almost the same.
> > there are some quirks to use the `celery_app.send_task()` as in debug mode i don't use the pool(so no `await` keyword), but in production we do (`await` keyword is needed). this can be fixed with some extra configuration, i guess.
> > @auvipy awesome news! i hope this time it will be merged and done in `v6` (i remember promises over `v5` inclusion)
> 
> I get very few money to work on celery right now. it is a huge overhaul work. take lots of time and technical difficulties

@auvipy I didn't see this response earlier. Would you be open to being sponsored to work on the feature. We have sponsored various Python and iOS project that make our core stack.

If you are please reach out to me on [Twitter](https://twitter.com/mdevraj) or our [corporate site](https://www.anomaly.ltd)

**Comment by tkauf15k** [comment]
since i worked recently on the same issue, my findings just as an additional comment... 
generally, i guess it's quite easy to come up with a workaround, either with _asgiref_ or just manually by spinning up an event loop. __asgiref__ ,though, i had issues with underlying dependencies like asyncpg. 
for me, __celery-pool-asyncio__  did not really work with celery 5.2.7 out of the box, due to a version conflict ("module 'celery.app.trace' has no attribute 'monotonic'"), so essentially, after https://github.com/celery/celery/commit/5b86b35c81ea5a1fbfd439861f4fee6813148d16 celery-pool-asyncio is broken. 

The main issue however, with these workarounds are tests. I have a few tests with a more "integrative" nature, with an asyncio test base (IsolatedAsyncioTestCase), where tasks are called eagerly. Then i basically have an async method invoking a synchronous one (`task.apply`) that in turn then wants to invoke an async one. Apparantely, you cannot run the async to completion in the existing loop, and scheduling a task is no option. 

My workaround just for the tests now is to spin up a thread and wait for it's completion within the "main-event loop" and spin up a new loop for the async task. Although i'm kinda unhappy with this dirty hack, it's the best i could come up with (cause it becomes event trickier when you have further dependencies, like asyncpg that bind to a specific loop etc..

**Comment by francipvb** [comment]
For a while, the asgiref sync pattern should work, but this is a non-optimized way.

I will keep an eye on this, too.

**Comment by avin-kavish** [comment]
I'm just looking to do parallel aiohttp requests within a single celery task. Does anyone know whether that's possible?

**Comment by lucasgadams** [comment]
Kind of crazy to not have asyncio support yet!

**Comment by Nusnus** [comment]
> Kind of crazy to not have asyncio support yet!

We're definitely on the same page my friend although as important as it is, as Celery is being currently used in many production environments, we're focusing our minimal resources on providing improved stability, bug fixes and still trying to push forward new features and so our pipeline is very thin and asyncio requires dedicated attention we currently do not have available.

We'd appreciate any additions the community can contribute but until then we have to adjust our priorities where it most hurts.

**Comment by stuaxo** [comment]
I wonder if by now async would be better to target than asyncio.
Celery as an async event loop implementation would be really interesting for instance.

**Comment by tkauf15k** [comment]
> I wonder if by now async would be better to target than asyncio. Celery as an async event loop implementation would be really interesting for instance.

could you elaborate? async vs asyncio?

for me this topic is primarily a matter of being able to share methods between truly aio code (e.g. db access in a web api) and my background workers. i do not expect any/much benefits of using aio in celery apart from that. my expected behaviour would anyway be that a single process/worker takes a single task at a time and runs it to completion - so apart from concurrent behavior inside the task (several concurrent api calls, db accesses etc), i do not see any potential betwern the tasks / or actually want to have the worker based isolation and this could also be done by some basic sync-async wrapper or am i missing something?

**Comment by auvipy** [comment]
> Kind of crazy to not have asyncio support yet!

Stop doing unprofessional comments. if you can't add anything useful in any project, better not to bugging around.

**Comment by auvipy** [comment]
Fore anyone interested, please check https://github.com/celery/celery/issues/7874 and give it a try at your dev environment.

**Comment by stuaxo** [comment]
@tkauf15k I'd have a look at the difference between asyncio and async.  Asyncio was the first attempt in python 3.x but then this was thrown away (in effect) when the async keyword was added, in general both are not compatible.

Back on topic check out #7874 @auvipy linked, from the description it seems very sensible so if you have the chance to try it and provide feedback please do.

**Comment by dmwyatt** [comment]
> 'd have a look at the difference between asyncio and async. Asyncio was the first attempt in python 3.x but then this was thrown away (in effect) when the async keyword was added, in general both are not compatible.

This is not *exactly* correct (which I think you were gesturing at when you said "in effect").  async/await are tangential to asyncio. They are ways of creating coroutines. asyncio is one of several libraries providing the infrastructure or foundation for running coroutines in an asynchronous environment.

Async/await focus on the syntax and structure of asynchronous code, making it more readable and expressive. On the other hand, asyncio handles the low-level details of event loops, task scheduling, and I/O operations.  Previous to the async/await syntax introduction you used generators with `yield from` which was not ideal because it was easy to confused what was a coroutine and what was not and also you couldn't do the equivalent of `async with` or `async for`.

async/await are basically syntactic sugar.

**Comment by leowonglaw** [comment]
I've investigated there is a compatible workaround using `asgiref` and `threads pool`, for example:

```python
@celery_app.task
def my_task() -> None:
    async_to_sync(my_async_function)()
```
Now, use threads pool to increase the _concurrency_ per worker:

`celery -A app worker --pool=threads --concurrency=150`

This suits well for IO intensive tasks. Due to python GIL I think either having A) 1 CPU per worker or B) 1 worker per CPU in the same server is the recommended configuration. The concurrency will depend on your tasks.
I've also tried different packages and execution pools, but threads worked the best for me.

**Comment by spawn-guy** [comment]
@leowonglaw i think we need, kinda, both. A thread per cpu/core and then be able to limit the number of simultaneous tasks per thread. And a support for running async tasks.
I think we can create an asyncio(or other) pool per thread and then run sync and async tasks inside

**Comment by Nusnus** [comment]
Hey everyone,

Thanks to @spawn-guy’s comment I noticed there’s also this issue which I didn’t update on the recent updates.
TL;DR I am researching about the effort to add native asyncio support, from Kombu’s lower layers up to Celery’s tasks.
I’ve already prepared a short POC with x15-20 performance improvement on publishing 10,000 messages to RabbitMQ.

For more info, please read from this [comment](https://github.com/celery/celery/issues/6552#issuecomment-2346045087) to the end of the issue.

> @auvipy I didn't see this response earlier. Would you be open to being sponsored to work on the feature. We have sponsored various Python and iOS project that make our core stack.
> 
> If you are please reach out to me on [Twitter](https://twitter.com/mdevraj) or our [corporate site](https://www.anomaly.ltd)

As discussed in my [comment](https://github.com/celery/celery/issues/6552#issuecomment-2346045087), I am looking for funding to work on the support immediately. Is the offer to discuss the possibility still on the table? @devraj

Thank you 🙏

### Example from the POC
Standard API runtime: 3m 6.47s
Code blow (same flow/logic): 17.556s
```python
from __future__ import annotations

import asyncio
import time

from kombu.asyncio.transport.pyamqp import Channel, Connection

async def benchmark_asyncio() -> None:
    print("Running benchmark_asyncio with artificial delay...")

    connection_params = {
        "host": "localhost",
        "port": 5672,
        "login": "guest",
        "password": "guest",
        "virtualhost": "/",
    }

    num_iterations = 10_000
    delay = 0.001  # Simulated I/O delay in seconds

    connection = Connection(**connection_params)
    await connection.connect()
    print("Connection established.")

    channel: Channel = await connection.channel()
    print("Channel opened.")

    queue_name = "benchmark_queue"
    queue = await channel.declare_queue(queue_name, durable=False, auto_delete=False)
    print(f"Queue '{queue_name}' declared.")

    await queue.purge()
    print(f"Queue '{queue_name}' purged.")

    exchange = channel.default_exchange

    start_time = time.perf_counter()

    async def publish_message(i):
        message_body = f"Benchmark Message {i}"
        message = channel.prepare_message(
            body=message_body,
            content_type="text/plain",
            content_encoding="utf-8",
        )

        await exchange.publish(
            message,
            routing_key=queue_name,
        )
        await asyncio.sleep(delay)

    tasks = [publish_message(i) for i in range(num_iterations)]

    await asyncio.gather(*tasks)

    end_time = time.perf_counter()

    total_time = end_time - start_time
    avg_time = total_time / num_iterations

    print(f"Total time for {num_iterations} iterations: {total_time:.4f} seconds")
    print(f"Average time per iteration: {avg_time:.6f} seconds")

    await connection.close()
    print("Connection closed.”)
```

**Comment by devraj** [comment]
> Hey everyone,
> 
> Thanks to @spawn-guy’s comment I noticed there’s also this issue which I didn’t update on the recent updates. TL;DR I am researching about the effort to add native asyncio support, from Kombu’s lower layers up to Celery’s tasks. I’ve already prepared a short POC with x15-20 performance improvement on publishing 10,000 messages to RabbitMQ.
> 
> For more info, please read from this [comment](https://github.com/celery/celery/issues/6552#issuecomment-2346045087) to the end of the issue.
> 
> > @auvipy I didn't see this response earlier. Would you be open to being sponsored to work on the feature. We have sponsored various Python and iOS project that make our core stack.
> > If you are please reach out to me on [Twitter](https://twitter.com/mdevraj) or our [corporate site](https://www.anomaly.ltd)
> 
> As discussed in my [comment](https://github.com/celery/celery/issues/6552#issuecomment-2346045087), I am looking for funding to work on the support immediately. Is the offer to discuss the possibility still on the table? @devraj
> 
> Thank you 🙏
>

Let me send you an email @Nusnus

**Comment by Nusnus** [comment]
> > Hey everyone,
> > Thanks to @spawn-guy’s comment I noticed there’s also this issue which I didn’t update on the recent updates. TL;DR I am researching about the effort to add native asyncio support, from Kombu’s lower layers up to Celery’s tasks. I’ve already prepared a short POC with x15-20 performance improvement on publishing 10,000 messages to RabbitMQ.
> > For more info, please read from this [comment](https://github.com/celery/celery/issues/6552#issuecomment-2346045087) to the end of the issue.
> > > @auvipy I didn't see this response earlier. Would you be open to being sponsored to work on the feature. We have sponsored various Python and iOS project that make our core stack.
> > > If you are please reach out to me on [Twitter](https://twitter.com/mdevraj) or our [corporate site](https://www.anomaly.ltd)
> > 
> > 
> > As discussed in my [comment](https://github.com/celery/celery/issues/6552#issuecomment-2346045087), I am looking for funding to work on the support immediately. Is the offer to discuss the possibility still on the table? @devraj
> > Thank you 🙏
> 
> Let me send you an email @Nusnus

Thank you very much @devraj.
Looking forward to discussing this further 🔥

**Comment by auvipy** [comment]
> Standard API runtime: 3m 6.47s
> Code blow (same flow/logic): 17.556s

did the PoC also worked with celery? or it is kombu only ATM?

**Comment by Nusnus** [comment]
> > Standard API runtime: 3m 6.47s
> > Code blow (same flow/logic): 17.556s
> 
> did the PoC also worked with celery? or it is kombu only ATM?

Pure Kombu.
To benefit from native asyncio, we need to start from the lowest layer.
That hard part is indeed Celery, but the better Kombu’s support for asyncio, the easier it will be with Celery.

This is also an opportunity to fix long-due issues in Kombu and show some love to the internals.

**Comment by spawn-guy** [comment]
> > > Standard API runtime: 3m 6.47s
> > > Code blow (same flow/logic): 17.556s
> > 
> > did the PoC also worked with celery? or it is kombu only ATM?
> 
> Pure Kombu.
> To benefit from native asyncio, we need to start from the lowest layer.
> That hard part is indeed Celery, but the better Kombu’s support for asyncio, the easier it will be with Celery.
> 
> This is also an opportunity to fix long-due issues in Kombu and show some love to the internals.

Shall we discuss things around my other PR? 😉 I wanna refractor the api a bit

**Comment by Sauraj1234** [comment]
Hi All, 

I am facing same problem (want to use asyncio with celery), i appreciate all of your hard work to maintain 
this project my current condition where i am publishing the message from pika to rabbitmq 
core base publisher 
now i want to use asyncio for consumer in celery ( i know celery does not have the feature right now )
but can anyone have the custom/alternative solution for this problem

**Comment by rkrell** [comment]
> Hi All,
> 
> I am facing same problem (want to use asyncio with celery), i appreciate all of your hard work to maintain this project my current condition where i am publishing the message from pika to rabbitmq core base publisher now i want to use asyncio for consumer in celery ( i know celery does not have the feature right now ) but can anyone have the custom/alternative solution for this problem

I use this workaround in celery 5.5.X:

```python
import asyncio
from inspect import isawaitable

import celery

class AsyncCeleryTask(celery.Task):
    abstract = True

    async def _run(self, *args, **kwargs):
        result = celery.Task.__call__(self, *args, **kwargs)
        if isawaitable(result):
            return await result
        return result

    def __call__(self, *args, **kwargs):
        return asyncio.run(self._run(*args, **kwargs))


class AsyncCelerySingleton(celery.Celery):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, task_cls=AsyncCeleryTask, **kwargs)


celery_app = AsyncCelerySingleton()
```

You may then provide mixed synchronous of asynchronous tasks, for example:

```python
@shared_task
async def my_async_task()
   ...

```

This works like a charm even though it is a hack, but a nice one :-)

**Comment by Sauraj1234** [comment]
First of all thanks you ;) for your reply  @rkrell 
With approach i have investigated with my code 
there is issue i am facing even if i implemented this approach it will block my 
event loop for asyncio
I am using sqlalchmey ORM for db query in worker when db insert query execute it will block the loop 
is this true ?  my use case FIFO with (ordered data) this sequential  approach
can above solution works for me for celery  ?

much appreciated  answer 

please let me know if i am doing something wrong 

Also interested in contributing opensource, tell me if you have any guide ( I am beginner in Opens source )

**Comment by spawn-guy** [comment]
@Sauraj1234 i am using https://github.com/the-wondersmith/celery-aio-pool with asyncpg and some code magic
